virtual_eeprom/virtual_eeprom.c

```diff
@@ -0,0 +1,219 @@
+    ret = save_eeprom_to_file(eeprom);
+    mutex_unlock(&eeprom->mutex);
+
+    return ret < 0 ? ret : 0;
+}
+
+static int save_eeprom_to_file(struct virtual_eeprom *eeprom)
+{
+    struct file *file;
+    loff_t pos = 0;
+    ssize_t ret;
+
+    file = filp_open(eeprom->bin_path, O_WRONLY | O_CREAT | O_TRUNC, 0644);
+    if (IS_ERR(file)) {
+        pr_err("Failed to open file %s for writing: %ld\n", eeprom->bin_path, PTR_ERR(file));
+        return PTR_ERR(file);
+    }
+
+    ret = kernel_write(file, eeprom->data, eeprom->size, &pos);
+    filp_close(file, NULL);
+
+    return ret < 0 ? ret : 0;
+}
+
+static int load_eeprom_from_file(struct virtual_eeprom *eeprom)
+{
+    struct file *file;
+    loff_t pos = 0;
+    ssize_t ret;
+
+    file = filp_open(eeprom->bin_path, O_RDONLY, 0644);
+    if (IS_ERR(file)) {
+	    // don't print read error if the file doesn't exist
+        if (PTR_ERR(file) != -ENOENT)
+            pr_err("Failed to open file %s for reading: %ld\n", eeprom->bin_path, PTR_ERR(file));
+        return PTR_ERR(file);
+    }
+
+    ret = kernel_read(file, eeprom->data, eeprom->size, &pos);
+    filp_close(file, NULL);
+
+    return ret < 0 ? ret : 0;
+}
+
+static int __init eeprom_init(void)
+{
+    struct device_node *np, *child;
+    int ret;
+
+    // Find the parent device node
+    np = of_find_node_by_name(NULL, "virtual_eeproms");
+    if (!np) {
+        pr_err("No device node found for virtual_eeproms\n");
+        return -ENODEV;
+    }
+
+    // Iterate through all child nodes
+    for_each_child_of_node(np, child) {
+        const char *name = child->name;
+        const char *bin_path;
+        u32 size;
+
+        if (of_property_read_u32(child, "size", &size)) {
+            pr_err("Failed to read size for %s\n", name);
+            continue;
+        }
+
+        if (of_property_read_string(child, "bin-path", &bin_path)) {
+            pr_err("Failed to read bin-path for %s\n", name);
+            continue;
+        }
+
+        struct virtual_eeprom *eeprom = kzalloc(sizeof(struct virtual_eeprom), GFP_KERNEL);
+        if (!eeprom) {
+            pr_err("Failed to allocate memory for %s\n", name);
+            continue;
+        }
+
+        eeprom->data = kzalloc(size, GFP_KERNEL);
+        if (!eeprom->data) {
+            pr_err("Failed to allocate memory for EEPROM %s\n", name);
+            kfree(eeprom);
+            continue;
+        }
+
+        mutex_init(&eeprom->mutex);
+        eeprom->size = size;
+        eeprom->name = name;
+        eeprom->bin_path = bin_path;
+
+        // Create unique platform device name by appending EEPROM name
+        eeprom->pdev = platform_device_register_simple(name, -1, NULL, 0);
+        if (IS_ERR(eeprom->pdev)) {
+            pr_err("Failed to register platform device for %s\n", name);
+            kfree(eeprom->data);
+            kfree(eeprom);
+            continue;
+        }
+
+        ret = load_eeprom_from_file(eeprom);
```

Comment:
Need a case where the file is smaller than the eeprom size specified in the device tree. Error/Increase file size?

Reply:
fixed, module prints  error and skips registering the eeprom which has size mismatch with dts